### ISSUE TITLE
Node simplify or deepEqual ordering

### ISSUE THREAD

**Issue by jcarolus** [josdejong/mathjs#1260]
The equals method sometimes returns false, even tough both expressions are equal, because the deepEqual method does not care about the order of `args` arrays. E.g.: `mathjs.simplify('a+x^2').equals(mathjs.simplify('x^2+a'))` returns `false`.

This is only an issue when comparing symbolic algebraic expressions, but definitely an interesting use case I would say.

The `simplify.rules` seems not to have a way to affect the ordering of the Node tree and it does not seem the right place to do this.
Changing the deepEqual to clone, order and then compare with strict order will impact performance, so not something to be done by default.
Maybe the Node object could be extended with an `equalsSymbolic` or optional argument for `equals` to cater for this use case?

Willing to create a PR for it and would like to know which is preferred, or perhaps another solution?

**Comment by josdejong** [comment]
That's an interesting case @jcarolus . 

I think the special cases that we have here is that the order of arguments for add `+` and multiply `*` don't change the mathematical meaning of it so can be considered equal. I think it's only these two cases right? Or are there more?

I think we can there are two different directions we can go:

1. Extend `simplify` so it orders the arguments somehow in an unambiguous way (like numbers first from low to high, symbols second alphabetically, then operators, then functions). Bascially we have to implement a `.compare` method to compare any type of node with any other type of node and use that in `simplify`.
2. Like you suggest we could implement an `equalsSymbolic` or pass an option to `equals` to ignore the ordering of arguments for both `add` and `multiply`.

I'm not sure yet which one of the two I like best.

**Comment by jcarolus** [comment]
@josdejong indeed, only `+` and `*` (not a mathematician myself, but I can not think of other cases).

Extending simplify with ordering could be a feature that helps with presentation as well - this could turn into a new `simplify.orderingRules` which can be used in the same way as `simplify.ordering`. 

For the second option, it would make sense to clone both arguments, so your Nodes won't change after you have compared them (I can imagine that would break all kinds of tests), but comes with a performance penalty - hence optional. 

Third option / starting option: add an `order` to `Node`, which uses `compare` to order the arguments.

**Comment by josdejong** [comment]
That's a good point. Right now a series of additions like `a + b + c` is parsed as two nested operators `add`, like `((a + b) + c)` or `add(add(a, b), c)`, we should replace that with a single operator `add` having three (or more) arguments, like `add(a, b, c)`. And the same for multiply. 

The underlying functions `add` and `multiply` already support this so it should be relatively easy to change this behavior.

**Comment by gwhitney** [comment]
Now that #2384 is merged, there is another possibility that works pretty well as of right now:
```
function equalsSymbolic(a:string, b:string) {
  return math.simplify(math.parse(`${a} - ${b}`)).toString() === '0';
}
```
It certainly gives the correct answer of true for `a*x^2` and `x^2*a`. 
(Is there a better way than ...`.toString() === 0` to test if an expression is exactly the constant 0 expression?)
I think this is the way other 'CAS's actually do this, by normalizing the difference of the two expressions.
Let me know if you'd like a PR that beefs this up and adds an equalsSymbolic function (and if so, if that's exactly the name you'd like).

**Comment by josdejong** [comment]
Performing `a - b = 0` on two expressions to see whether they are equal is quite a nifty solution 😄. Yes please give it a try Glen!

About the naming: currently we have `equal`, `deepEqual`, `equalText`, so maybe we can call this either `symbolicEqual` or `equalSymbolic`. At least, a singular name to keep it consistent with the other equal functions. Does that make sense?

**Comment by gwhitney** [comment]
Well, it seems there are two possibilities here, which have potentially different implications for naming:

1) It seems to me that closest to the originally posted question/suggestion would be a method of Node which returns whether the Node it is called from is symbolically equal to another Node; this would be a loosening of the current Node method `equals`, and so in this case `symbolicEquals` seems like it would be appropriate, or

2) A top-level `math.xxx` function that can take two parse trees and return whether they are symbolically equal; that would seem most similar to `math.deepEqual` and so should likely be named `math.symbolicEqual`.

Let me know which of these you feel fits better into mathjs and I will put this on the list of PRs to work on.

**Comment by josdejong** [comment]
Good point. I hadn't thought about `Node.equals`. Where mathjs came from was just a bunch of functions, but we more and more have classes with methods too. I want to be sparse with adding methods to classes, since the classes can become so large and pull in a lot of dependencies, resulting in less opportunity for tree shaking to make a small bundle with just the things that you actually use. Since this new function will depend on `resolve`, it makes sense to me to implement it as standalone function, preventing a circular dependency and adding this big dependency to the core class `Node`. Does that make sense?

**Comment by josdejong** [comment]
Thanks Glen!

**Comment by josdejong** [comment]
Published in `v10.3.0`
